### DISCUSSION TITLE
Moebius shader code to node

### DISCUSSION THREAD

**Discussion by JPLost** [bnpr/Malt#124] [Q&A] [answered]
You had example of Moebius like shader in code.
Turning this to node caused me to think how do you define the shadow color?
Everything behind diffused surface is shadowed by default.

[Moebius_Nodes.zip](https://github.com/bnpr/Malt/files/7163931/Moebius_Nodes.zip)

If I understand correctly, the idea is to use NPR_lit_surface. Calculate the spots, where there is shadow and bypass them by feeding this information to PixelOutput id channel?

**Comment by pragma37** [comment, marked as answer]
The get_line functions can detect lines between ID boundaries.
By default, each object has its own ID so it detects lines between objects.

So, for getting lines at the boundaries between light and shadows you want to assign different ids to areas in light and areas in shadows.
There are many ways to implement this depending on what you want to achieve. For example:

![imagen](https://user-images.githubusercontent.com/36610999/133451118-9810eebe-37cd-4856-a746-3acc47bf22fb.png)
[Moebius_Nodes.zip](https://github.com/bnpr/Malt/files/7170849/Moebius_Nodes.zip)

The problem with setting IDs like that is that they are also used for detecting self shadows, so you may want to set the ID only in the PRE_PASS.
There's no way to detect the current pass directly from nodes at the moment, but you could add this function to your custom node library:
```glsl
bool is_prepass()
{
    #ifdef PRE_PASS
    {
        return true
    }
    #else
    {
        return false;
    }
    #endif
}
```

**Reply by JPLost** [reply]
Oh wow, you learn something new every day. 👍 

So it looks like the get_diffuse_gradient has been converted to black-white float value between 0-1.
Then you have linked Surface.id to the inline code and defined a value 10 as int.
As the gradient is set to constant, it should give three id's that are 1, 2 and 3?

Is there any handy reference site for these functions?

**Reply by pragma37** [reply]
> As the gradient is set to constant, it should give three id's that are 1, 2 and 3?

Kind of, the ID is offset by the luma value. So in this case, the ID can go from the new ID value, to the new ID value + 10.
The value (b) will change how many ID "steps" each object has.
You can replace the gradient with a black to white one to easily visualize it.

> Is there any handy reference site for these functions?

Not really. I'd like to make docs with example images, but that's unlikely to happen until the 1.0 release.

**Comment by JPLost** [comment]
[Moebius_Nodes_Revised.zip](https://github.com/bnpr/Malt/files/7190405/Moebius_Nodes_Revised.zip)

Lets add the updated version with is_prepass node added. 
Requires newest development version release to work.
